crates/rpc/rpc/src/trace.rs

```diff
@@ -549,6 +479,84 @@ where
         }
         traces
     }
+
+    /// Helper method to fetch blocks for trace filtering
+    async fn get_blocks_for_trace_range(
+        &self,
+        start: u64,
+        end: u64,
+    ) -> Result<Vec<RecoveredBlock<<Eth::Provider as BlockReader>::Block>>, Eth::Error> {
+        let end_block_hash = match self.provider().block_hash(end) {
+            Ok(Some(hash)) => hash,
+            _ => {
+                // If we can't get the end block hash, use the original implementation
+                return self
+                    .provider()
+                    .recovered_block_range(start..=end)
+                    .map_err(Eth::Error::from_eth_err)
+            }
+        };
+
+        let cached_blocks = match self
+            .inner
+            .eth_api
+            .cache()
+            .get_cached_parent_blocks(end_block_hash, (end - start + 1) as usize)
+            .await
+        {
+            Some(blocks) => blocks,
+            None => {
+                return self
+                    .provider()
+                    .recovered_block_range(start..=end)
+                    .map_err(Eth::Error::from_eth_err)
+            }
+        };
+
+        let oldest_cached_block_number =
+            cached_blocks.last().map(|b| b.header().number()).unwrap_or(0);
+
+        if oldest_cached_block_number <= start && cached_blocks.len() as u64 >= (end - start + 1) {
+            // We have all blocks needed in the cache!
+            // Filter and reverse to get them in ascending order
+            let mut result: Vec<_> = cached_blocks
+                .into_iter()
+                .filter(|b| {
+                    let num = b.header().number();
+                    num >= start && num <= end
+                })
+                .map(|arc_block| arc_block.as_ref().clone()) // Convert Arc<RecoveredBlock> to RecoveredBlock
+                .collect();
+            result.reverse();
+            return Ok(result);
+        }
+
+        // We need to fetch missing blocks from start to the oldest cached block - 1
+        if oldest_cached_block_number > start {
+            let missing_blocks = self
+                .provider()
+                .recovered_block_range(start..=(oldest_cached_block_number - 1))
+                .map_err(Eth::Error::from_eth_err)?;
+
+            // Filter and reverse cached blocks to get the ones we need in ascending order
+            let mut cached_result: Vec<_> = cached_blocks
+                .into_iter()
+                .filter(|b| {
+                    let num = b.header().number();
+                    num >= start && num <= end
```

Comment:
hmm, not exactly sure we need all of this. 

I'm okay with only hitting the cache for end blocks close to latest

and then the only thing that we'd need is frontfilling the missing range to start, so this looks a bit too complex